Sources/NIOSSL/SSLConnection.swift

```diff
@@ -31,6 +31,7 @@ internal let sslConnectionExDataIndex = CNIOBoringSSL_SSL_get_ex_new_index(0, ni
 /// handle the "must wait for more data" case by calling it out directly.
 enum AsyncOperationResult<T> {
     case incomplete
+    case lookup
```

Comment:
I don't think this is needed: the existing `incomplete` flows should work fine, shouldn't they?

Reply:
I don't think they do. There are multiple cases where `incomplete` branch will be called, and then you end up in a state where you have to track if you added a complete handler to the future already. Seemed much cleaner to have an explicit branch for lookups. The lookup case felt important enough to me to not hide those details

Reply:
We can attach the completion handler inside the callback itself though right?

Reply:
If you attach complete handler inside the callback then there's no way to signal back out to `SSLConnection` that it should continue the handshake. You could resume the handshake inside the complete handler (which I originally had) but then `SSLConnection` doesn't know about it and never fires the complete events.

Reply:
You can follow the pattern we use for the custom certificate verification code: https://github.com/apple/swift-nio-ssl/blob/7c381eb6083542b124a6c18fae742f55001dc2b5/Sources/NIOSSL/SSLCallbacks.swift#L240-L255.

Reply:
I think the cert_cb registration still gets the `SSLConnection` as something it can pull out, so you can probably still do this from where it is.

Reply:
I made a mistake in my last couple comments, its not `SSLConnection` that we need access to in order to complete the handshake, it's `SSLHandler` that needs to call `doHandshakeStep` which in turns takes result from `SSLConnection.doHandshake` and fires the necessary events.